src/main/java/org/openjdk/jextract/impl/UnsupportedFilter.java

```diff
@@ -210,66 +188,72 @@ private boolean checkFunctionTypeSupported(Declaration decl, Type.Function func,
         return true;
     }
 
-    private static final Type.Visitor<String, Void> UNSUPPORTED_VISITOR = new Type.Visitor<>() {
+    private static final Type.Visitor<Type, Boolean> UNSUPPORTED_VISITOR = new Type.Visitor<>() {
         @Override
-        public String visitPrimitive(Type.Primitive t, Void unused) {
-            Optional<MemoryLayout> layout = Type.layoutFor(t);
-            if (layout.isPresent() && layout.get() instanceof PaddingLayout) {
-                return t.kind().typeName();
-            } else {
-                return null;
-            }
+        public Type visitPrimitive(Type.Primitive t, Boolean allowVoid) {
+            return switch (t.kind()) {
+                case Char16, Float128, HalfFloat, Int128, WChar -> t;
+                case LongDouble -> TypeImpl.IS_WINDOWS ? null : t;
+                case Void -> allowVoid ? null : t;
+                default -> null;
+            };
         }
 
         @Override
-        public String visitFunction(Type.Function t, Void unused) {
+        public Type visitFunction(Type.Function t, Boolean allowVoid) {
             for (Type arg : t.argumentTypes()) {
-                String unsupported = firstUnsupportedType(arg);
+                Type unsupported = firstUnsupportedType(arg, false);
                 if (unsupported != null) {
                     return unsupported;
                 }
             }
-            String unsupported = firstUnsupportedType(t.returnType());
+            Type unsupported = firstUnsupportedType(t.returnType(), true);
             if (unsupported != null) {
                 return unsupported;
             }
             return null;
         }
 
         @Override
-        public String visitDeclared(Type.Declared t, Void unused) {
-            for (Declaration d : t.tree().members()) {
-                if (d instanceof Declaration.Variable variable) {
-                    String unsupported = firstUnsupportedType(variable.type());
-                    if (unsupported != null) {
-                        return unsupported;
-                    }
+        public Type visitDeclared(Type.Declared t, Boolean allowVoid) {
+            if (t.tree().kind() == Kind.STRUCT || t.tree().kind() == Kind.UNION) {
+                if (!isValidStructOrUnion(t.tree())) {
+                    return t;
                 }
             }
             return null;
         }
 
+        private boolean isValidStructOrUnion(Scoped scoped) {
```

Comment:
This is a method which checks whether a struct or union is valid. Basically the conditions are:
* its size should be available (e.g. not an opaque struct)
* if it's an anonymous struct, it should have a known first offset (otherwise, the anon struct should be just skipped)